### ISSUE TITLE
Reactive ecosystem for Rust

### ISSUE THREAD

**Issue by aepsil0n** [reactive-rs/meta#1]
Dear community,

I have been hacking on [Carboxyl](https://github.com/aepsil0n/carboxyl) for quite a while, initially for video game projects @Moredread and I used to work on. In any case, I would very much love to build on this idea in the future. There seems to be some potential to build an ecosystem of reactive libraries & applications based on and around carboxyl. This would help people willing to build performance-sensitive reactive applications in a functional manner with Rust.

This thread is intended to serve as a platform for generating and discussing initial ideas about what kinds of libraries or tools would be helpful. Feel free to share what kind of applications you want to build with this.

also pinging anyone, who may be interested in this:
@doppioslash @killercup @miosta @mitchmindtree @Moredread

**Comment by doppioslash** [comment]
Hello!

I'm invested in FRP, my day job is in Elm and I'm working on it being viable to use in games.
But I think Rust could be a great platform to have performant and type safe FRP running on.
It's a much easier sell to gamedevs as well.
I'm still learning Rust, but I mean to try implementing a Time Travelling Debugger like Elm's ( demo here https://youtu.be/vS3yzUo7l8Y?t=23s ) in Rust.

So my main use case is games, especially 3d games. And testing how far a TTD can go.

**Comment by jj-jabb** [comment]
The one thing that prevents me from considering Carboxyl is the LGPL license - especially since importing a crate statically links it - I'm sure there's a way to build it separately and dynamically link it but that seems like a lot of friction.

Would you consider switching to the Mozilla Public License?

**Comment by tupshin** [comment]
I'm working on some projects in this space as well. Things that I've noted that would be beneficial would be 
1. static guarantees of interior immutability
2. static guarantees of functional purity
3. Serializable lambda expresions https://github.com/rust-lang/rfcs/issues/1022
4.Tuple and struct field iteration (mostly to aid in working with compile time code generation
5. type level laziness
5. HKTs
6. Dependent types

:)

**Comment by tupshin** [comment]
And yes, for any ecosystem of libraries in the rust ecosystem to really take off, there will have to be a standard practice of dual Apache/MPL licensing, in addition to any other license that somebody might use.

**Comment by aepsil0n** [comment]
Cool to see, that there is some interest in this. ;)

@doppioslash: That's a pretty cool idea. I know the TTD feature in Elm, but your idea makes me wonder how much of this is possible in Rust.

@tupshin all these things seem desirable. It all kind of feels out of scope for this project though, as it mostly concerns the Rust language itself, does it not?

@jj-jabb @tupshin: This license conversation keeps popping up. I am very much an advocate of free software, so I am hesitant to just swap over to Apache/MPL. My interpretation of the LGPL would be that you only have to provide a way to dynamically link the library for proprietary software built with it, you do not have to do that by default. This is the reason, why I compromised by switching from GPL to LGPL. The MPL is something I would at least take into consideration, if it is that big of a deal. Personally I do not necessarily want to cater to the interests of people building proprietary software. Dual-licensing with Apache, however, would definitely defeat the purpose of protecting user rights.

**Comment by bjz** [comment]
If you are setting your sights on game development, then licensing is a really big deal if you have aspirations beyond small hobby projects. We'd just end up with another library for reactive programming with a more permissive license, and a more fractured ecosystem.

**Comment by dylanede** [comment]
I'm actually also in the process of writing a combination of libraries for FRP in Rust. My approach is heavily inspired by Elm, specifically Evan Czaplicki's thesis on concurrent FRP. Based on that paper I've managed to make my FRP implementation have the same semantics and it's very nearly lock-free (just waiting on a good implementation of a lock-free multi-producer multi-consumer queue). It also load balances across CPU cores. I haven't released it yet as I am dogfooding it first to test the API. I am currently working on the fundamentals of a GUI library suitable for use with FRP in Rust. My first public contribution from this is [RustType](https://github.com/dylanede/rusttype). Now I'm working on an automatic constrained layout library, which is really crucial for ease of use. I've implemented the Cassowary algorithm but I'm not happy with its shortcomings so I'm trying a different approach now. I plan to release all of this under dual MIT/Apache.

**Comment by potocpav** [comment]
I am building a glium+carboxyl reactive GUI library similar to Blender (splittable panels and all). The progress has been slow however, in part because the semantics of Carboxyl are not documented precisely enough. Considered writing Elerea clone, but then decided that all the monads would get very hairy in Rust.

**Comment by mitchmindtree** [comment]
@dylanede sounds cool! Have you seen the [crossbeam implementation of the Michael Scott Queue](http://aturon.github.io/crossbeam-doc/crossbeam/sync/struct.MsQueue.html)? It's MPMC.

**Comment by vvanders** [comment]
To echo on what @bjz said you won't get much traction inside the games space without a permissive license. That industry is much more interested in building games than open source concerns. For them techniques rather than libraries tend to be what's shared openly.

**Comment by Moredread** [comment]
@tupshin What is the idea behind Apache/MPL dual licensing? AFAIK the MPL is more restrictive than the Apache license, so why would you not choose to comply to the Apache license? Or are the attribution clauses such a large problem?

**Comment by jacmoe** [comment]
I understand - and admire - Free Software.
However, for a programming library like this, I prefer MIT.
It is GPL-compatible.
And I think that it would mean much larger adoption. :)

**Comment by dylanede** [comment]
@mitchmindtree That looks like just the thing I need! Last time I checked Crossbeam I didn't spot that. Hmm. Mind you that was a very long time ago. I've managed to get everything else in the library lock free other than the MPMC queue implementation, so this should be the final thing to make it truly lock-free.

@jacmoe That's exactly the reason I'm using permissive licenses. I really want everything and anything Rust related to take off as much as possible (makes it more likely for me to get a job in Rust in the future), and restrictive licenses can only inhibit that.

**Comment by kerinin** [comment]
Just to jump on the bandwagon, I also wrote a library based on the Elm concurrent FRP paper: https://github.com/kerinin/cfrp-rs

**Comment by tupshin** [comment]
Regarding licensing, it is worth looking at [Mozilla licensing page](https://wiki.mozilla.org/License_Policy/Mozilla_Project_Licensing) and sections 7 and 8 of [this](https://doc.rust-lang.org/1.4.0/complement-project-faq.html).

Personally, I will be following rust's dual licensing policy, and therefore won't be able to depend on any crates that are not. Additionally, as I will necessarily be implementing a lot of frp patterns with a more permissive license, I can't even read your code, as it is currently stands.  

Regarding my choice to require such permissive licensing for myself, it is largely driven by a desire to enhance, augment, and interoperate with the apache big data and distributed systems ecosystem.

**Comment by skade** [comment]
An alternative to LGPL is the GPL with linking exception as libgit2: https://github.com/libgit2/libgit2#libgit2---the-git-linkable-library

It especially avoids problems where the LGPL imposes restrictions on the shipped product (mainly, that relinking the product with a different version must be possible, which is - for example - not possible with signed applications on the Apple App Store).

**Comment by aepsil0n** [comment]
Somewhat surprised that the licensing issue appears to be this important to people. From my point of view, this is a political question and not merely a technical or legal one. Nonetheless, I do not really see value in getting into a deep discussion about it in this thread. So I would like to address the practical ramifications.

To clarify my position on that: my intention is not to keep developers of proprietary software from using the library, but to make them contribute bug fixes and improvements upstream. If MPL or GPL with linking exception turns out to better achieve that goal, I am happy to switch. However, it is not my primary goal to get big industry to use my code. If they do adopt it, fine, but I am not going to preemptively conform to their needs. This is what using Apache or MIT means to me to be clear.

Would moving to MPL or GPL linking exception address the concerns voiced by people here? By that, I mean: *Are you willing to contribute to the proposed ecosystem, when its core component is licensed like that?*

**Comment by jj-jabb** [comment]
I am willing to contribute to a ecosystem where the core component is licensed under the MPL, and I think that the MPL achieves the goals you are looking for, where changes to the software must be published while at the same time allowing static linking, which plays nice with the cargo ecosystem. Furthermore it is an official unmodified license which (to me) is more appealing than a modified license.

**Comment by Moredread** [comment]
I'm wondering a bit what your goal is when lobbying for a permissive license like MIT or Apache?

Growth and a large developer base on it's own doesn't seem to me worth much, if you don't consider selfish reasons like better employability.

I don't see much worth in winning developers of proprietary software as users, when they don't contribute back to the ecosystem. According to my observations most vendors will not engage much with the community or contribute much less than they profit from using free software, but it seems worse in communities where permissive licenses are common. E.g. most web startups don't make their webapps free software, while using high quality free software web frameworks or JS packages.

 At least there is some impetus (or coercion ^^)  to contribute back when a copy-left license is used.

On the other hand I'm a bit disturbed by the strong call for a more permissive license. I'm open to discuss this issue, but IMO it is on the people that want a license change to make a good argument for it, and that I think is missing here, as I think "a broader audience" is a very weak one.
Also the more or less implicit thread to boycott the project when no change of the license happens is IMHO rather hostile, esp. in the context of a discussion of founding a new ecosystem when it is not qualified with any further argument and posed as an ultimatum. :(

**Comment by doppioslash** [comment]
If you want a lib to be used in games, indie or AAA, use a permissive license.

Gamedevs are not necessarily aware, or interested about open source licensing.
Tbh it's hard enough to get them to consider seriously FRP without adding license stubbornness in the mix.

If you don't care about that, then whatever.
But my motivation is making better gamedev tools so I'll probably either roll my own or use another FRP lib with more permissive licensing.

**Comment by doppioslash** [comment]
@aepsil0n I think a Rust TTD, especially if limited to an FRP ecosystem is most likely possible :)

**Comment by tupshin** [comment]
> To clarify my position on that: my intention is not to keep developers of proprietary software from using the library, but to make them contribute bug fixes and improvements upstream.

To clarify my own position, my intention is to ensure that the code I work with can be used by developers of **permissively** open source software. That means no dependency on less permissive software. Again, that is my personal choice, and means that I can't rely on anything not similarly permissively licensed. I am not asking anybody else to change their licensing, merely stating as unavoidable fact, that if there is significant [L]GPL FRP work done in Rust, then there will necessarily be an additional FRP Rust ecosystem that I will be contributing to. 

Now this does create a directional imbalance...you will be free to use my code, but I will not be free to use yours, but that is fine, if that is the choice that you want to make.

**Comment by jacmoe** [comment]
> I don't see much worth in winning developers of proprietary software as users, when they don't contribute back to the ecosystem.

Developers of proprietary software is regular people, and they contribute at least as much as everyone else does.

They might contribute more than non-proprietary users simple because they are getting paid to work full time writing software.

> At least there is some impetus (or coercion ^^) to contribute back when a copy-left license is used.

That is naive. No, licensing has nothing to do with whether or not people are contributing.

People contribute because of practical reasons. Because it is much better to have changes/improvements contributed back upstream to avoid having to juggle / keep up / maintain a stack of patches.

A permissive license allows that to happen much more easily.

A copy-left license will not mean that people contribute back because the license says that they must do so. No. It simply means that they won't contribute back at all, or that they will seek other (and more permissively licensed) alternatives.

> On the other hand I'm a bit disturbed by the strong call for a more permissive license. I'm open to discuss this issue, but IMO it is on the people that want a license change to make a good argument for it, and that I think is missing here, as I think "a broader audience" is a very weak one.

Rust is dual licensed under Apache and MIT, so the burden of argumentation falls on you guys. ;)
Because going against the practice of the entire Rust eco-system requires a strong argument.

**Comment by potocpav** [comment]
> Developers of proprietary software is regular people, and they contribute at least as much as everyone else does.

My experience differs. The code written by a developer for a company is owned by that company, and the developer typically has to push very hard to make the management let him contribute back.

> People contribute because of practical reasons. Because it is much better to have changes/improvements contributed back upstream to avoid having to juggle / keep up / maintain a stack of patches.

In practice, it has always been easier for me to juggle a stack of commits. It can be a real effort to get major changes merged upstream (backwards incompatibility, not useful to the majority of users, etc.)

> Rust is dual licensed under Apache and MIT, so the burden of argumentation falls on you guys. ;)
Because going against the practice of the entire Rust eco-system requires a strong argument.

Keeping current license requires just doing nothing. So unless a compelling argument is made, it's quite likely to stay the same :)

**Comment by aepsil0n** [comment]
Okay, my takeaway is opening aepsil0n/carboxyl#100, i.e. trying to relicense Carboxyl under MPL 2.0, because you managed to convince me that LGPL a practical impediment to collaboration within the community that I had not intended. While I have not seen a compelling argument that using Apache/MIT is imperative, I am open to a heterogenous licensing landscape within this ecosystem.

However, I will not take part in any further licensing discussion here, because I don't want to derail the discussion any more.

**Comment by aepsil0n** [comment]
@potocpav You mentioned issues with Carboxyl's semantics not being precise enough. It would be really helpful to get more concrete feedback on that, so that it can be addressed. I only have a vague idea in the back of my mind, what you could be referring to (such as dynamic switching not being referentially transparent). Would you mind opening an issue with more details on the Carboxyl repository?

**Comment by potocpav** [comment]
@aepsil0n Will try to do that. I'm sorry for being so vague.

**Comment by aepsil0n** [comment]
@potocpav No worries. Thank you for pointing it out in the first place ;)
